### PR TITLE
Update HL7 Validation Error Messaging

### LINKED ISSUE #5643: Update the API error msg description returned
## Problem statement

As a user, I want to see the error message describe clearly what is expected in the data fields identified below, so that I can correct the data error and resend data in a format that will be accepted by the ReportStream API.

Refer - https://cdc.sharepoint.com/:x:/r/teams/USDSatCDC/_layouts/15/Doc.aspx?sourcedoc=%7B3fac1cd7-3045-479b-b46e-1f2550a3415e%7D&action=editnew&cid=598df61b-8bdd-4e59-ac83-2bae83b41783

Scope - where column A is Ticket#5643

## What you need to know

Initial doc from O&O: https://cdc.sharepoint.com/teams/USDSatCDC/_layouts/15/doc.aspx?sourcedoc={6327b3dc-2ace-4757-aa62-ab8dcfffb535}&action=edit

Doc of Errors & Warning Info from Seth: https://cdc.sharepoint.com/teams/USDSatCDC/Shared%20Documents/Forms/AllItems.aspx?id=%2Fteams%2FUSDSatCDC%2FShared%20Documents%2FPRIME%20ReportStream%2FTeam%20Documents%2FExperience%20Team%2FEngineering%2FMisc%20Experience%20Eng%20Docs%2FReportStream%20Errors%20%26%20Warnings%20Inventory%20%2D%20sample%20files&viewid=17db3ff6%2D95c5%2D4562%2Da1f3%2D12b5097b5947

Link to Error & Warnings Inventory SAMPLE FILES from Seth: https://cdc.sharepoint.com/teams/USDSatCDC/Shared%20Documents/Forms/AllItems.aspx?id=%2Fteams%2FUSDSatCDC%2FShared%20Documents%2FPRIME%20ReportStream%2FTeam%20Documents%2FExperience%20Team%2FEngineering%2FMisc%20Experience%20Eng%20Docs%2FReportStream%20Errors%20%26%20Warnings%20Inventory%20%2D%20sample%20files&viewid=17db3ff6%2D95c5%2D4562%2Da1f3%2D12b5097b5947

Research from #6307: https://cdc.sharepoint.com/:x:/r/teams/USDSatCDC/_layouts/15/Doc.aspx?sourcedoc=%7B3fac1cd7-3045-479b-b46e-1f2550a3415e%7D&action=editnew&cid=598df61b-8bdd-4e59-ac83-2bae83b41783

## Acceptance criteria
- [ ] As recommended in Column I,  https://cdc.sharepoint.com/:x:/r/teams/USDSatCDC/_layouts/15/Doc.aspx?sourcedoc=%7B3fac1cd7-3045-479b-b46e-1f2550a3415e%7D&action=editnew&cid=598df61b-8bdd-4e59-ac83-2bae83b41783
Scope - where column A is Ticket#5643

- [ ] Modify existing unit tests as needed

## Excluded
- Any new tests for uncovered conditions

### PR DESCRIPTION
**Pull request by kevinhaube**
This PR uses the document Seth created from #6307 to update the back-end code with new error reporting messages. Please use the utilities included in the attached ticket #5643 to test this on your local machine.

Test Steps:
1. Pull latest on this branch and spin up the Docker containers, RS Router, and Postman
2. Use Postman to send a message to `http://localhost:7071/api/validate` with any of the provided error-laden files
3. Ensure the messages presented match the desired messages in the spreadsheet from Seth and O&O.

## Changes
- Updates to error reporting messages for file validation

### Testing
- [X] Tested locally?
- [X] Ran `./prime test` or `./gradlew testSmoke` against local Docker ReportStream container?
- [ ] (For Changes to /frontend-react/...) Ran `npm run lint:write`? 
- [ ] Added tests?

### Process
- [ ] Are there licensing issues with any new dependencies introduced?
- [ ] Includes a summary of what a code reviewer should test/verify?
- [ ] Updated the release notes?
- [ ] Database changes are submitted as a separate PR?
- [ ] DevOps team has been notified if PR requires ops support?

## Linked Issues
- Fixes #5643 

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?

### CODE REVIEW

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 1]
```diff
@@ -365,7 +389,14 @@ class Hl7Serializer(
                 itemIndex = messageIndex
             )
         } catch (e: Exception) {
-            val msg = "${e.localizedMessage} ${e.stackTraceToString()}"
+            val msg = when (e) {
+                // Couldn't use InvalidPhoneMessage because it requires params we
+                // cannot access from this scope
+                is NumberParseException ->
+                    "The string supplied is not a valid phone number. Reformat to a 10-digit phone number (e.g. (555)" +
+                        " 555-5555)."
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:397]: This is not the only format we accept for phone numbers. We also accept international numbers if they contain the country code. Your message should reflect that.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt [hunk 2]
```diff
@@ -268,9 +277,20 @@ class Hl7Serializer(
...
+                    )
+                )
             } else {
-                errors.add(InvalidHL7Message("Error parsing HL7 message: ${e.localizedMessage}"))
+                errors.add(
+                    InvalidHL7Message(
+                        "Error parsing HL7 message. Primitive " +
+                            "value '08112020' requires to be empty or a HL7 datetime string " +
+                            "at PID-29."
+                    )
```
Comment (MauriceReeves-usds) [inline prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:292]: This error message is very specific, especially for such a generic throw location. Also `08112020` is a valid value for PID-29 (Patient Date of Death). What's the story about this message?

**Review by MauriceReeves-usds — CHANGES_REQUESTED** [review]
There are a few things that need to be addressed. Please ping me if you have any questions. Thanks

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt (hunk 1 above)
Comment (kevinhaube) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:397]: Please refer to the pre-approved error messages work done in tandem between Seth and the O&O team: https://cdc.sharepoint.com/:x:/r/teams/USDSatCDC/_layouts/15/Doc.aspx?sourcedoc=%7B3fac1cd7-3045-479b-b46e-1f2550a3415e%7D&action=editnew&cid=598df61b-8bdd-4e59-ac83-2bae83b41783

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt (hunk 2 above)
Comment (kevinhaube) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:286]: Looked into this further. This _is_ where the error is caught and converted to an error message for RS, but the tricky part is that this is parsing every field, and the error thrown can't be caught anywhere between the library handling parsing and the RS app.

`arrayOf("ORU", "R01") contentEquals msgType -> parser.parse(cleanedMessage)`

This is the line that throws it from the `ca.uhn.hl7v2.validation.impl.RuleSupport` class (part of Hapi)

That said, the difference in the messages are as follows:
Before

> Error parsing HL7 message: ca.uhn.hl7v2.validation.ValidationException: Validation failed: Primitive value '08112022' requires to be empty or a HL7 datetime string at PID-29(0)

After

> Error parsing HL7 message. Primitive value '08112020' requires to be empty or a HL7 datetime string at PID-29.

I'll pass this back to the O&O team: Do we need to make any changes on this presentation or not? Seems like we'd only be stripping things that sound too techincal.

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt (hunk 1 above)
Comment (kevinhaube) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:392]: Update: From O&O sync on Wed, Oct 5

We will forego updating this error message entirely for this ticket, and wait for approved messaging on the full details of telephone number support! I will remove this line and the error message will remain as follows for now:

> The string supplied did not seem to be a phone number. Error type: NOT_A_NUMBER. The string supplied did not seem to be a phone number. at com.google.i18n.phonen

File: prime-router/src/main/kotlin/serializers/Hl7Serializer.kt (hunk 2 above)
Comment (kevinhaube) [reply prime-router/src/main/kotlin/serializers/Hl7Serializer.kt:286]: Update: From O&O sync on Wed, Oct 5

Stripping out the excess text `ca.uhn.hl7v2.validation.ValidationException: Validation failed:` from the above error is considered low priority by the O&O team and will not be included in this revision of the error messages.

**Pull request merged**